core/TypePtr.h

```diff
@@ -73,30 +73,27 @@ class TypePtr final {
     };
     tagged_storage store;
 
-    // Top bit indicates if value is inlined into pointer.
-    // We use a 0 in top bit to indicate not inlined so that nullptr (which has counter value 0) is naturally viewed as
+    // We use a 0 to indicate not inlined so that nullptr (which has counter value 0) is naturally viewed as
     // 'inlined'.
-    static constexpr tagged_storage NOT_INLINED_MASK = 0x8000000000000000;
-    static constexpr tagged_storage TAG_MASK = 0x7FFF000000000007;
+    static constexpr tagged_storage NOT_INLINED_MASK = 0x100;
+    // We use a 8-bit tag mask so that tags can be efficiently extracted via
+    // movzx on x86.
+    static constexpr tagged_storage TAG_MASK = 0xFF;
 
     static constexpr tagged_storage PTR_MASK = ~(NOT_INLINED_MASK | TAG_MASK);
 
     static tagged_storage tagToMask(Tag tag) {
-        auto val = static_cast<tagged_storage>(tag);
-        if (val >= 8) {
-            // Store the tag in the upper 16 bits of the pointer, as it won't fit in the lower three bits.
-            val <<= 48;
-        }
-        return val;
+        // Store the tag in the lower bits of the pointer, regardless of size.
+        return static_cast<tagged_storage>(tag);
     }
 
     static tagged_storage tagValue(Tag tag, u4 inlinedValue) {
         auto val = tagToMask(tag);
 
         // Store value into val.
-        val |= static_cast<tagged_storage>(inlinedValue) << 3;
+        val |= static_cast<tagged_storage>(inlinedValue) << 32;
```

Comment:
Why store the inlined value in the upper 32 bits? Are there more efficient instructions if it's in the upper 32-bits vs shifted by 16? Could you add a comment here explaining the reasoning?

Reply:
No strong reason, but storing it in the upper 32 implies that retrieving it only requires a shift and no masking (since the right shift by 32 implicitly masks).  I'll add a comment.